### ISSUE TITLE
Quaternionf.normalize needs precision improvement

### ISSUE THREAD

**Issue by lukehutch** [JOML-CI/JOML#213]
Given the quaternion

```
q = new Quaternionf(-0.001600552f, 3.4311073E-4f, 4.991863E-4f, 0.9999985f)
```

the code for `Quaternionf.normalize`, which is

```java
    public Quaternionf normalize() {
        float invNorm = (float) (1.0 / Math.sqrt(x * x + y * y + z * z + w * w));
        x *= invNorm;
        y *= invNorm;
        z *= invNorm;
        w *= invNorm;
        return this;
    }
```

calculates `invNorm == 1.0f`. Consequently, `q.normalize()` is idempotent, but should not be, because `Math.sqrt(x * x + y * y + z * z + w * w)` actually evaluates to `0.9999999701976772`.

Normalization should really produce a quaternion `(0, 0, 0, 1)` for this example, if `q` really is at the limit of precision for a `Quaternionf`.

I'm not sure of the right way to fix this, but with a value of `w` so close to `1.0f`, the value of `Math.toDegrees(q.angle())` should be close to `0.0`, but is actually `0.19782343594174043`, which is not insignificant. This is problematic when the angle is used to detect zero-rotations. Anything close to the unit quaternion (within a precision limit) should return something close to `0.0` for the rotation. (I say "something close to" because using some sort of precision epsilon to decide when to return exactly `0.0` would not give smooth results.)

**Comment by XenoAmess** [comment]
1.
cannot reproduce.
I write a test code as your description.
```
package com.xenoamess.cyan_potion.base;

import org.joml.Quaternionf;
import org.junit.jupiter.api.Test;

public class JomlTest {
    @Test
    public void test() {
        float x = -0.001600552F;
        float y = 3.4311073E-4F;
        float z = 4.991863E-4F;
        float w = 0.9999985F;
        Quaternionf q1 = new Quaternionf(x, y, z, w);
        q1.normalize(q1);
        float degree1 = (float) Math.sqrt(q1.angle());
        System.out.println(degree1);

        float invNorm = (float) (1.0 / Math.sqrt(x * x + y * y + z * z + w * w));
        x *= invNorm;
        y *= invNorm;
        z *= invNorm;
        w *= invNorm;
        System.out.println(x);
        System.out.println(y);
        System.out.println(z);
        System.out.println(w);
        Quaternionf q2 = new Quaternionf(x, y, z, w);
        float degree2 = (float) Math.sqrt(q2.angle());
        System.out.println(degree2);
    }
}
```
result is :

```
0.05754724
-0.001600552
3.4311073E-4
4.991863E-4
0.9999985
0.05875943
```
0.057 is small enough.

2. 
you should NEVER use any sort of precision epsilon to decide when to return exactly 0.0 in float/double format, but SHOULD write your own detecter.(a small range)
even 0.19 is small enough in usual cases.
If you need higher precision you need rewrite JOML using BigDecimal.
But I really don't think a game developer will need such a high precision.

**Comment by lukehutch** [comment]
Thanks for writing the testcase.

> you should NEVER use any sort of precision epsilon to decide when to return exactly 0.0 in float/double format

Yes, I know this. I was getting a situation with lots of quaternions near (0 0 0 1) were simply not changing after normalization, even though they should have been changing slightly. I traced it in the debugger to `(float) (1.0 / Math.sqrt(x * x + y * y + z * z + w * w))` evaluating to exactly `1.0f` when `x * x + y * y + z * z + w * w` was not `1.0f`. The act of taking the `sqrt` and reciprocal, and casting to `float`, cumulatively produced a loss of precision. I filed this bug report hoping that there might be a higher-precision way of calculating the normalization constant.

Rendering the numbers into decimal for the example I gave also gives rise to precision errors, since binary fractions cannot be represented exactly in decimal, hence your code can't reproduce it. I'm not sure how to grab an example in such a way that it can be reliably reproduced.

I guess I'll switch to using `Quaterniond` in the hopes of preserving as much precision as possible.
